crates/telio-starcast/src/nat.rs

```diff
@@ -0,0 +1,794 @@
+use crate::utils::{checksum, MutableIpPacket};
+use pnet_packet::{
+    ip::IpNextHeaderProtocols, ipv4::MutableIpv4Packet, ipv6::MutableIpv6Packet,
+    udp::MutableUdpPacket,
+};
+use std::{
+    hash::{DefaultHasher, Hash, Hasher},
+    net::{IpAddr, Ipv4Addr, Ipv6Addr, SocketAddr},
+    time::Duration,
+};
+use telio_utils::{telio_log_warn, LruCache};
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    /// The packet is too short to be a valid IP packet
+    PacketTooShort,
+    /// The packet has an unexpected IP version
+    UnexpectedIpVersion,
+    /// The packet has an unexpected transport protocol
+    UnexpectedTransportProtocol,
+    /// Cannot map the packet
+    CannotMap,
+    /// There is no mapping for the given port
+    MappingNotFound,
+}
+
+pub trait Nat {
+    /// Translate incomming packet (from the transport socket to the multicast peer)
+    /// Change the source to the multicast peer's ip and natted port
+    fn translate_incomming(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+    /// Translate outgoing packet (from the multicast peer to the transport socket)
+    /// Change the destination to the peer's original ip and port
+    fn translate_outgoing(&mut self, packet: &mut [u8]) -> Result<(), Error>;
+}
+
+type Mapper = fn(src_addr: SocketAddr) -> u16;
+
+fn default_mapper(src_addr: SocketAddr) -> u16 {
+    let mut hasher = DefaultHasher::new();
+    match src_addr.ip() {
+        IpAddr::V4(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+        IpAddr::V6(src_ip) => Hash::hash_slice(&src_ip.octets(), &mut hasher),
+    };
+    src_addr.port().hash(&mut hasher);
+    hasher.finish() as u16 // TODO: think if we want to use the full 16 bits or maybe some range e.g. >=1024
+}
+
+pub struct StarcastNat {
+    mappings: LruCache<u16, SocketAddr>,
+    vpeer_ipv4_addr: Ipv4Addr,
+    vpeer_ipv6_addr: Ipv6Addr,
+    mapper: Mapper,
+}
+
+impl StarcastNat {
+    /// Create new starcast nat with default capcity of 2048 entries and 2 minute timeout
+    /// 2min default is taken from (https://datatracker.ietf.org/doc/html/rfc4787#section-4.3)
+    pub fn new(vpeer_ipv4: Ipv4Addr, vpeer_ipv6: Ipv6Addr) -> Self {
+        const DEFAULT_CACHE_CAPACITY: usize = 2048;
```

Comment:
NIT: move contants to the top of the file

Reply:
@packgron I'm not a fan TBH and I wouldn't do it